### PR TITLE
Inject packets in bulk

### PR DESCRIPTION
**Pull request by nurlybekovnt**
Bulk packet injection was added.

### CODE REVIEW

**Comment by yerden** [comment]
I should have been come up with this earlier but this implementation implies copying packet data which may be expensive. Maybe we should pass `C.uintptr_t` but we should be aware of GC.

File: snf/inject.h [hunk 1]
```diff
@@ -9,20 +9,17 @@
 
 static inline int
 snf_inject_send_bulk(snf_inject_t inj, int timeout_ms, int flags,
-                     const uint8_t *pkts, uint32_t n_pkts, const uint32_t *lengths)
+                     const void **pkts, uint32_t n_pkts, const uint32_t *lengths)
```
Comment (yerden) [inline snf/inject.h:12]: shouldn't it be `void * const *pkts`?

File: snf/inject.go [hunk 2]
```diff
@@ -254,15 +257,16 @@ func (s *Sender) SendBulk(pkts [][]byte) error {
 		return err
 	}
 
+	s.guardPkts = pkts
 	s.pkts = s.pkts[:0]
 	s.len = s.len[:0]
 	for _, pkt := range pkts {
-		s.pkts = append(s.pkts, pkt...)
+		s.pkts = append(s.pkts, uintptr(unsafe.Pointer(&pkt[0])))
```
Comment (yerden) [inline snf/inject.go:264]: `&pkts[i]`?

File: snf/inject.go [hunk 3]
```diff
@@ -148,8 +148,11 @@ type Sender struct {
 	frags []C.struct_snf_pkt_fragment
 
 	// buffers for injecting in bulk
-	pkts []byte
+	pkts []uintptr
 	len  []uint32
+
+	// protect the memory from GC
+	guardPkts [][]byte
```
Comment (yerden) [inline snf/inject.go:155]: just a note, this is valid if Sender is allocated in heap.

File: snf/inject.go [hunk 4]
```diff
@@ -254,15 +257,16 @@ func (s *Sender) SendBulk(pkts [][]byte) error {
...
 	for _, pkt := range pkts {
-		s.pkts = append(s.pkts, pkt...)
+		s.pkts = append(s.pkts, uintptr(unsafe.Pointer(&pkt[0])))
 		s.len = append(s.len, uint32(len(pkt)))
 	}
 
-	return retErr(C.snf_inject_send_bulk(injHandle(s.InjectHandle), s.timeoutMs,
-		s.flags, (*C.uint8_t)(&s.pkts[0]), C.uint32_t(len(s.pkts)), (*C.uint32_t)(&s.len[0])))
+	return retErr(C.snf_inject_send_bulk(injHandle(s.InjectHandle), s.timeoutMs, s.flags,
+		(*unsafe.Pointer)(unsafe.Pointer(&s.pkts[0])), C.uint32_t(len(s.pkts)), (*C.uint32_t)(&s.len[0])))
```
Comment (yerden) [inline snf/inject.go:269]: you can just pass `unsafe.Pointer(&s.pkts[0])` and handle it as `uintptr_t *pkts` in C code. Looks simpler.

File: snf/inject.go (hunk 2 above)
Comment (nurlybekovnt) [reply snf/inject.go:264]: My bad, sorry

Comment (nurlybekovnt) [reply snf/inject.go:264]: `s.pkts` has pointers to byte arrays, which are raw packets. `&pkts[i]` is a pointer to a slice that has a pointer to a byte array, which can be obtained by slicing, so in this case, it should be `&pkts[i][0]` equivalent to `&pkt[0]`.

File: snf/inject.go (hunk 4 above)
Comment (nurlybekovnt) [reply snf/inject.go:269]: You mean `(*C.uintptr_t)(unsafe.Pointer(&s.pkts[0]))`?

File: snf/inject.h (hunk 1 above)
Comment (nurlybekovnt) [reply snf/inject.h:12]: I changed it to `uintptr_t *pkts` as you recommended below.

File: snf/inject.go [hunk 5]
```diff
@@ -143,6 +146,13 @@ type Sender struct {
 
 	// fragment buffer
 	frags []C.struct_snf_pkt_fragment
+
+	// buffers for injecting in bulk
+	pkts []uintptr
+	len  []uint32
```
Comment (yerden) [inline snf/inject.go:152]: `C.uint32_t` would be better then

File: snf/inject.go [hunk 6]
```diff
@@ -240,6 +250,25 @@ func (s *Sender) Send(pkt []byte) error {
...
+	if err := s.checkSignal(); err != nil {
+		return err
+	}
+
+	s.guardPkts = pkts
+	s.pkts = s.pkts[:0]
+	s.len = s.len[:0]
+	for _, pkt := range pkts {
+		s.pkts = append(s.pkts, uintptr(unsafe.Pointer(&pkt[0])))
+		s.len = append(s.len, uint32(len(pkt)))
```
Comment (yerden) [inline snf/inject.go:265]: `C.uint32_t`?

File: snf/inject.go [hunk 7]
```diff
@@ -143,6 +146,13 @@ type Sender struct {
 
 	// fragment buffer
 	frags []C.struct_snf_pkt_fragment
+
+	// buffers for injecting in bulk
+	pkts []uintptr
```
Comment (yerden) [inline snf/inject.go:151]: `C.uintptr_t`

File: snf/inject.go (hunk 4 above)
Comment (yerden) [reply snf/inject.go:269]: if you follow the `C.uintptr_t` declaration then it's just `&s.pkts[0]`.

File: snf/inject.go (hunk 5 above)
Comment (nurlybekovnt) [reply snf/inject.go:152]: Fixed

File: snf/inject.go (hunk 6 above)
Comment (nurlybekovnt) [reply snf/inject.go:265]: Fixed

File: snf/inject.go (hunk 7 above)
Comment (nurlybekovnt) [reply snf/inject.go:151]: Fixed

File: snf/inject.go (hunk 4 above)
Comment (nurlybekovnt) [reply snf/inject.go:269]: Fixed

File: snf/inject.go (hunk 3 above)
Comment (nurlybekovnt) [reply snf/inject.go:155]: Added the note in documentation

File: snf/inject.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
+#ifndef _INJECT_H_
+#define _INJECT_H_
+
+#include <stdint.h>
+
+#ifndef USE_MOCKUP
+#include <snf.h>
+#endif
+
+static inline int
```
Comment (yerden) [inline snf/inject.h:10]: you may return `struct compound_int` as in `ring_recv_many` in `wrapper.h`. Then use `intErr` to return both number of sent packets and error code if any.

```
#include "wrapper.h"
...
statc inline struct compound_int
snf_inject_send_bulk(snf_inject_t inj, int timeout_ms, int flags,
                     uintptr_t *pkts, uint32_t n_pkts, const uint32_t *lengths)
{
    struct compound_int x;
    int i;

    for (i = 0; i < n_pkts; i++)
    {
        x.rc = snf_inject_send(inj, timeout_ms, flags, (void *)pkts[i], lengths[i]);
        if (x.rc)
        {
            break;
        }
    }

    x.i[0] = i;
    return x;
}
```

Comment (nurlybekovnt) [reply snf/inject.h]: Thanks for the help. Used your recommendations.

**Pull request merged**